modules/OCPP201/OCPP201.cpp

```diff
@@ -840,18 +841,50 @@ void OCPP201::ready() {
+                // transform request forward to libocpp
+                ocpp::v201::Get15118EVCertificateRequest ocpp_request;
+                ocpp_request.exiRequest = certificate_request.exiRequest;
+                ocpp_request.iso15118SchemaVersion = certificate_request.iso15118SchemaVersion;
+                ocpp_request.action = [](const types::iso15118_charger::CertificateActionEnum& action) {
+                    switch (action) {
+                    case types::iso15118_charger::CertificateActionEnum::Install:
+                        return ocpp::v201::CertificateActionEnum::Install;
+                    case types::iso15118_charger::CertificateActionEnum::Update:
+                        return ocpp::v201::CertificateActionEnum::Update;
+                    }
+                }(certificate_request.certificateAction);
+
+                auto ocpp_response = this->charge_point->on_get_15118_ev_certificate_request(ocpp_request);
+                EVLOG_info << "received response: " << ocpp_response;
+                // transform response, inject action, send to associated EvseManager
+                const auto everest_response_status = [](const ocpp::v201::Iso15118EVCertificateStatusEnum& action) {
+                    switch (action) {
+                    case ocpp::v201::Iso15118EVCertificateStatusEnum::Accepted:
+                        return types::iso15118_charger::Status::Accepted;
+                    case ocpp::v201::Iso15118EVCertificateStatusEnum::Failed:
+                        return types::iso15118_charger::Status::Failed;
+                    }
+                }(ocpp_response.status);
+                const types::iso15118_charger::Response_Exi_Stream_Status everest_response{
+                    everest_response_status, certificate_request.certificateAction, ocpp_response.exiResponse};
+                EVLOG_info << "return response: " << everest_response;
+                this->r_evse_manager.at(evse_id - 1)->call_set_get_certificate_response(everest_response);
+            });
 
-        r_system->subscribe_log_status([this](types::system::LogStatus status) {
-            this->charge_point->on_log_status_notification(get_upload_log_status_enum(status.log_status),
-                                                           status.request_id);
-        });
+        evse_id++;
     }
+    r_system->subscribe_firmware_update_status([this](const types::system::FirmwareUpdateStatus status) {
+        this->charge_point->on_firmware_update_status_notification(
+            status.request_id, get_firmware_status_notification(status.firmware_update_status));
+    });
+
+    r_system->subscribe_log_status([this](types::system::LogStatus status) {
+        this->charge_point->on_log_status_notification(get_upload_log_status_enum(status.log_status),
+                                                       status.request_id);
+    });
```

Comment:
Question: These subscriptions don't seem related to M01-M05? It's probably not relevant to the review, but what functionality do they fulfill?

Reply:
hmm no this was just a minor "bug" (not even really) that is discovered: the r_system connection is independent of the evse, but due to an error the system subscriptions were included in the loop above (and thus done redundantly multiple times); but please verify this again in the code!

Reply:
Hm, from what I can tell by glancing through the framework code, it seems that subscribing to the same var multiple times will cause EVerest to subscribe to the MQTT topic multiple times, there doesn't seem to be a check to prevent that. It's possible that the MQTT client simply uses the last subscription it's given, so this wouldn't be a bug, but I'd agree this is probably more like what we want.

Is it possible to exit early out of the loop over all EVSEs above without crashing?